ImperatorToCK3/Source/Mappers/ReligionMapper/ReligionMapping.cpp

```diff
@@ -1,22 +1,98 @@
 #include "ParserHelpers.h"
+#include "Log.h"
+#include "Mappers/RegionMapper/ImperatorRegionMapper.h"
+#include "Mappers/RegionMapper/CK3RegionMapper.h"
 
 mappers::ReligionMapping::ReligionMapping(std::istream& theStream)
 {
 	registerKeyword("ck3", [this](const std::string& unused, std::istream& theStream) {
-		const commonItems::singleString religionString(theStream);
-		ck3Religion = religionString.getString();
+		ck3Religion = commonItems::singleString{ theStream }.getString();
 	});
 	registerKeyword("imp", [this](const std::string& unused, std::istream& theStream) {
-		const commonItems::singleString religionString(theStream);
-		impReligions.insert(religionString.getString());
+		impReligions.insert(commonItems::singleString{ theStream }.getString());
 	});
-	registerKeyword("region", [this](const std::string& unused, std::istream& theStream) {
-		const commonItems::singleString regionString(theStream);
-		ck3Regions.insert(regionString.getString());
+	registerKeyword("ck3Region", [this](const std::string& unused, std::istream& theStream) {
+		ck3Regions.insert(commonItems::singleString{ theStream }.getString());
+	});
+	registerKeyword("impRegion", [this](const std::string& unused, std::istream& theStream) {
+		imperatorRegions.insert(commonItems::singleString{ theStream }.getString());
+	});
+	registerKeyword("ck3Province", [this](const std::string& unused, std::istream& theStream) {
+		ck3Provinces.insert(commonItems::singleULlong{ theStream }.getULlong());
+	});
+	registerKeyword("impProvince", [this](const std::string& unused, std::istream& theStream) {
+		imperatorProvinces.insert(commonItems::singleULlong{ theStream }.getULlong());
 	});
 	registerRegex(commonItems::catchallRegex, commonItems::ignoreItem);
 
 	parseStream(theStream);
 	clearRegisteredKeywords();
-}
\ No newline at end of file
+}
+
+std::optional<std::string> mappers::ReligionMapping::religionMatch(const std::string& impReligion,
+                                                                   const unsigned long long ck3ProvinceID, const unsigned long long impProvinceID) const
+{
+	// We need at least a viable Imperator religion
+	if (impReligion.empty())
+		return std::nullopt;
+
+	if (!impReligions.contains(impReligion))
+		return std::nullopt;
+
+	// This is a straight province check, not regions.
+	if (!ck3Provinces.empty() && ck3Regions.empty())
+		if (!ck3Provinces.contains(ck3ProvinceID))
+			return std::nullopt;
+	if (!imperatorProvinces.empty() && imperatorRegions.empty())
+		if (!imperatorProvinces.contains(impProvinceID))
+			return std::nullopt;
+
+	// Asking for a regions check without an incoming province is pointless.
+	if (!ck3Regions.empty() && !ck3ProvinceID)
+		return std::nullopt;
+	if (!imperatorRegions.empty() && !impProvinceID)
+		return std::nullopt;
+
+	// This is a CK3 regions check, that checks if a provided province is within that CK3 region.
+	if (!ck3Regions.empty() || !imperatorRegions.empty())
+	{
+		if (!ck3RegionMapper)
```

Comment:
no need to check for these on every single match. Check when initing the seek in the mapper before iterating over mappings.

Reply:
Yeah, it's better.
You might want to do it like this in CK3ToEU4's CultureMapper, which I based the change on.